From 6d3e78af5a7ac575dce9a35ddb918b7440d5aaeb Mon Sep 17 00:00:00 2001
From: "he, bo" <bo.he@intel.com>
Date: Tue, 18 Sep 2012 13:21:43 +0800
Subject: [PATCH 51/78] wl12xx: flush the workqueue during __ieee80211_suspend

BZ: 57560

Add the quiesce interface to flush the workqueue in
__ieee80211_suspend.

Change-Id: Iecbf975553e5e6c89d8691fc63da9b2dd322eaf0
Signed-off-by: Yanmin Zhang <yanmin.zhang@intel.com>
Signed-off-by: he, bo <bo.he@intel.com>
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Reviewed-on: http://android.intel.com:8080/74181
Reviewed-by: lab_cactus <lab_cactus@intel.com>
Reviewed-by: Casasnovas, QuentinX <quentinx.casasnovas@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: Jacquet, CyrilX <cyrilx.jacquet@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/78008
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |   19 +++++++++++++++++++
 wl12xx-compat/include/net/mac80211.h             |    1 +
 wl12xx-compat/net/mac80211/pm.c                  |    3 +++
 3 files changed, 23 insertions(+), 0 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 66633ad..51dbcae 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -2306,6 +2306,24 @@ static int wl1271_op_suspend(struct ieee80211_hw *hw,
 	return 0;
 }
 
+static int wl1271_op_quiesce(struct ieee80211_hw *hw)
+{
+	struct wl1271 *wl = hw->priv;
+
+	flush_delayed_work_sync(&wl->scan_complete_work);
+	flush_work_sync(&wl->tx_work);
+	struct wl12xx_vif *wlvif;
+	wl12xx_for_each_wlvif(wl, wlvif) {
+		del_timer_sync(&wlvif->rx_streaming_timer);
+		flush_work_sync(&wlvif->rx_streaming_enable_work);
+		flush_work_sync(&wlvif->rx_streaming_disable_work);
+	}
+
+	flush_delayed_work_sync(&wl->elp_work);
+
+	return 0;
+}
+
 static int wl1271_op_resume(struct ieee80211_hw *hw)
 {
 	struct wl1271 *wl = hw->priv;
@@ -5720,6 +5738,7 @@ static const struct ieee80211_ops wl1271_ops = {
 #ifdef CONFIG_PM
 	.suspend = wl1271_op_suspend,
 	.resume = wl1271_op_resume,
+	.quiesce = wl1271_op_quiesce,
 #endif
 	.config = wl1271_op_config,
 	.prepare_multicast = wl1271_op_prepare_multicast,
diff --git a/wl12xx-compat/include/net/mac80211.h b/wl12xx-compat/include/net/mac80211.h
index 15d9cc2..cbf8852 100644
--- a/wl12xx-compat/include/net/mac80211.h
+++ b/wl12xx-compat/include/net/mac80211.h
@@ -2144,6 +2144,7 @@ struct ieee80211_ops {
 #ifdef CONFIG_PM
 	int (*suspend)(struct ieee80211_hw *hw, struct cfg80211_wowlan *wowlan);
 	int (*resume)(struct ieee80211_hw *hw);
+	int (*quiesce)(struct ieee80211_hw *hw);
 #endif
 	int (*add_interface)(struct ieee80211_hw *hw,
 			     struct ieee80211_vif *vif);
diff --git a/wl12xx-compat/net/mac80211/pm.c b/wl12xx-compat/net/mac80211/pm.c
index 8160b97..f74265a 100644
--- a/wl12xx-compat/net/mac80211/pm.c
+++ b/wl12xx-compat/net/mac80211/pm.c
@@ -148,6 +148,9 @@ int __ieee80211_suspend(struct ieee80211_hw *hw, struct cfg80211_wowlan *wowlan)
 	/* need suspended to be visible before quiescing is false */
 	barrier();
 	local->quiescing = false;
+	if (local->ops->quiesce)
+		local->ops->quiesce(&local->hw);
+	flush_workqueue(local->workqueue);
 
 	return 0;
 }
-- 
1.7.9

